-gsource-map does not work on Windows if username contains non-ASCII Unicode characters.

If username has non-ASCII Unicode characters, then on Windows the `TEMP` environment variable will point to a path that has Unicode chars in it (e.g. `TEMP=C:\Users\Matti Meikäläinen\AppData\Local\Temp`)

In such a mode, using `-gsource-map` command line flag fails on Windows.

e.g.

```
C:\ä ö\emsdk\emscripten\main>set TEMP
TEMP=C:\ä ö\temp

C:\ä ö\emsdk\emscripten\main>test\runner other.test_wasm_sourcemap_relative_paths
Running test_other: (1 tests)
test_wasm_sourcemap_relative_paths (test_other.other.test_wasm_sourcemap_relative_paths) ... a.cpp a.cpp
   a.cpp
     same CWD for compiler and linker
FAIL

======================================================================
FAIL: test_wasm_sourcemap_relative_paths (test_other.other.test_wasm_sourcemap_relative_paths)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\ä ö\emsdk\emscripten\main\test\test_other.py", line 10781, in test_wasm_sourcemap_relative_paths
    test('a.cpp')
    ~~~~^^^^^^^^^
  File "C:\ä ö\emsdk\emscripten\main\test\test_other.py", line 10774, in test
    self.assertIn('"%s"' % expected_source_map_path, read_file('a.out.wasm.map'))
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: '"a.cpp"' not found in '{"version":3,"sources":["../../../../../../303/244 /303/266/emsdk/emscripten/main/out/test/a.cpp","/emsdk/emscripten/system/lib/compiler-rt/stack_ops.S","/emsdk/emscripten/system/lib/compiler-rt/stack_limits.S","/emsdk/emscripten/system/lib/pthread/library_pthread_stub.c","/emsdk/emscripten/system/lib/libc/musl/src/stdio/ofl.c","/emsdk/emscripten/system/lib/libc/musl/src/stdio/__lockfile.c","/emsdk/emscripten/system/lib/libc/musl/src/stdio/fflush.c"],"sourcesContent":[],"names":[],"mappings":"6XAOA,iCACE,QAAA,kBCQF,GACA,MACA,KAKA,MAEA,EAEA,CAEA,EACA,CACA,EACA,MACA,EACA,EAIA,OACA,ECWA,OAEA,MAMA,MAGA,EACA,CACA,EACA,CACA,MAEA,EAYA,OACA,MACA,CACA,EAnDA,OACA,EAIA,OACA,GCmVwB,GAEE,GChXzB,YACA,MAAA,GAKA,YACD,ECbA,GAaC,GASD,MCjBK,UAEC,cAA2B,SAAP,aACpB,UAA2B,SAAP,MAAH,MAEb,QAAV,GAAE,KACC,oBAAA,GAAA,QAEA,IADO,OAAW,KAAN,GAAkB,QAAH,MAC3B,eAHH,KAAE,MAKA,MAyBF,IApBC,kBAAA,GAAA,QAkBA,IAfO,WAAW,KAAN,SACX,iBACQ,KAAJ,MACH,QAMK,OAAW,OAAN,KAA8B,OAAP,GAAX,0BAGL,OAAU,OACrB,GAER,cAED,EAAA"}'

----------------------------------------------------------------------
Ran 1 test in 0.533s

FAILED (failures=1)

C:\ä ö\emsdk\emscripten\main>
```

It looks like the source map generation has generated an incorrect source path: 

```
"../../../../../../303/244 /303/266/emsdk/emscripten/main/out/test/a.cpp"
```
when it was expected to have resolved to just `a.cpp`.

If the `TEMP` directory points to a path without Unicode chars, then the test passes on Windows.

--- a/test/test_other.py
+++ b/test/test_other.py
@@ -10757,7 +10757,7 @@ int main() {
     ensure_dir('build')
 
     def test(infile, source_map_added_dir=''):
-      expected_source_map_path = 'a.cpp'
+      expected_source_map_path = 'A ä☃ö Z.cpp'
       if source_map_added_dir:
         expected_source_map_path = source_map_added_dir + '/' + expected_source_map_path
       print(infile, expected_source_map_path)
@@ -10778,10 +10778,10 @@ int main() {
         self.run_process([EMCC, 'a.o', '-gsource-map'], cwd='build')
         self.assertIn('"../%s"' % expected_source_map_path, read_file('build/a.out.wasm.map'))
 
-    test('a.cpp')
+    test('A ä☃ö Z.cpp')
 
-    ensure_dir('inner')
-    test('inner/a.cpp', 'inner')
+    ensure_dir('inner Z ö☃ä A')
+    test('inner Z ö☃ä A/A ä☃ö Z.cpp', 'inner Z ö☃ä A')
 
   def test_wasm_sourcemap_extract_comp_dir_map(self):
     wasm_sourcemap = importlib.import_module('tools.wasm-sourcemap')

--- a/tools/wasm-sourcemap.py
+++ b/tools/wasm-sourcemap.py
@@ -197,6 +197,25 @@ def remove_dead_entries(entries):
     block_start = cur_entry
 
 
+# Given a string that has non-ASCII UTF-8 bytes 128-255 stored as octal sequences (\200 - \377), decode
+# the sequences back to UTF-8. E.g. "C:\\\303\244 \303\266\\emsdk\\emscripten\\main" -> "C:\\ä ö\\emsdk\\emscripten\\main"
+def decode_octal_encoded_utf8(str):
+  out = bytearray(len(str))
+  i = 0
+  o = 0
+  final_length = len(str)
+  while i < len(str):
+    if str[i] == '\\' and (str[i+1] == '2' or str[i+1] == '3'):
+      out[o] = int(str[i+1:i+4], 8)
+      i += 4
+      final_length -= 3
+    else:
+      out[o] = ord(str[i])
+      i += 1
+    o += 1
+  return out[:final_length].decode('utf-8')
+
+
 def extract_comp_dir_map(text):
   map_stmt_list_to_comp_dir = {}
   chunks = re.split(r"0x[0-9a-f]*: DW_TAG_compile_unit", text)
@@ -205,7 +224,7 @@ def extract_comp_dir_map(text):
     if stmt_list_match is not None:
       stmt_list = stmt_list_match.group(1)
       comp_dir_match = re.search(r"DW_AT_comp_dir\s+\(\"([^\"]+)\"\)", chunk)
-      comp_dir = comp_dir_match.group(1) if comp_dir_match is not None else ''
+      comp_dir = decode_octal_encoded_utf8(comp_dir_match.group(1)) if comp_dir_match is not None else ''
       map_stmt_list_to_comp_dir[stmt_list] = comp_dir
   return map_stmt_list_to_comp_dir
 
@@ -251,12 +270,12 @@ def read_dwarf_entries(wasm, options):
 
     include_directories = {'0': comp_dir}
     for dir in re.finditer(r"include_directories\[\s*(\d+)\] = \"([^\"]*)", line_chunk):
-      include_directories[dir.group(1)] = os.path.join(comp_dir, dir.group(2))
+      include_directories[dir.group(1)] = os.path.join(comp_dir, decode_octal_encoded_utf8(dir.group(2)))
 
     files = {}
     for file in re.finditer(r"file_names\[\s*(\d+)\]:\s+name: \"([^\"]*)\"\s+dir_index: (\d+)", line_chunk):
       dir = include_directories[file.group(3)]
-      file_path = os.path.join(dir, file.group(2))
+      file_path = os.path.join(dir, decode_octal_encoded_utf8(file.group(2)))
       files[file.group(1)] = file_path
 
     for line in re.finditer(r"\n0x([0-9a-f]+)\s+(\d+)\s+(\d+)\s+(\d+)(.*?end_sequence)?", line_chunk):
@@ -352,8 +371,8 @@ def main():
 
   logger.debug('Saving to %s' % options.output)
   map = build_sourcemap(entries, code_section_offset, options)
-  with open(options.output, 'w') as outfile:
-    json.dump(map, outfile, separators=(',', ':'))
+  with open(options.output, 'w', encoding='utf-8') as outfile:
+    json.dump(map, outfile, separators=(',', ':'), ensure_ascii=False)
 
   if options.strip:
     wasm = strip_debug_sections(wasm)

--- a/tools/wasm-sourcemap.py
+++ b/tools/wasm-sourcemap.py
@@ -205,8 +205,8 @@ def decode_octal_encoded_utf8(str):
   o = 0
   final_length = len(str)
   while i < len(str):
-    if str[i] == '\\' and (str[i+1] == '2' or str[i+1] == '3'):
-      out[o] = int(str[i+1:i+4], 8)
+    if str[i] == '\\' and (str[i + 1] == '2' or str[i + 1] == '3'):
+      out[o] = int(str[i + 1:i + 4], 8)
       i += 4
       final_length -= 3
     else:

